Internally use MB as unit to request new resources in SLURM site adapter

giffels opened the pull request:
This pull request change the internal used unit to request memory using the SLURM site adapter to Megabytes. Externally memory is still handled in Gigabytes in `TARDIS`. This was necessary since the SLURM workload manager does not accept floating point variable to its `--mem` argument as reported in #196.

Currently the following command is used to submit a drone requesting 2.5 GB memory in the SLURM workload manager:
```
sbatch --mem=2.5gb ...
``` 
which fails.

After the update the following command is used:

```
sbatch --mem=2500mb
```
which will succeed.

This pull request changes only the internal behaviour and not the external handling of memory requests in `TARDIS` and it fixes #196.

HerrHorizontal reviewed (dismissed):
Looks fine! I have nothing to complain about.

olifre commented:
Thanks, that looks great!
I took the time to check the Slurm sources, and indeed:
https://github.com/SchedMD/slurm/blob/3e7b49a16cc7403245a2d8d7bd1f2b0df3228128/src/common/slurm_protocol_defs.c#L6022
* `gb` uses 1000 multipliers
* `g` and `gib` use `1024` multipliers

So the factor also does not change existing behaviour ;-).

RHofsaess commented:
> Thanks, that looks great!
> I took the time to check the Slurm sources, and indeed:
> https://github.com/SchedMD/slurm/blob/3e7b49a16cc7403245a2d8d7bd1f2b0df3228128/src/common/slurm_protocol_defs.c#L6022
> 
> * `gb` uses 1000 multipliers
> * `g` and `gib` use `1024` multipliers
> 
> So the factor also does not change existing behaviour ;-).

Perfect, thanks. This would have been my question!
Nothing else to add from my side.

giffels commented:
I will add additional unittests and a short comment to the code. Will ask for re-reviewing soon. Thanks!

maxfischer2781 commented:
Just wondering: Why not go for bytes directly? As far as I can tell, SLURM supports that as well.

Bytes never can be fractional, and it frees maintainers from having to look up whether `mb` is MB or MiB.

HerrHorizontal commented on tardis/adapters/sites/slurm.py:
```diff
@@ -174,7 +174,9 @@ def sbatch_cmdline_options(self, drone_uuid, machine_meta_data_translation_mappi
-            f"TardisDrone{key}={value}"
+            f"TardisDrone{key}={int(value)}"
+            if isinstance(value, float)
+            else f"TardisDrone{key}={value}"
```
Can `value` be any non-numerical data type? If no, then this distinction between `float` and other numerical data-types might not be needed.

maxfischer2781 replied:
It might be easier to always convert it to ``int``, no matter the type.

giffels replied:
Since it can be a string, that would fail.

giffels replied:
To be more precise, a string not convertible to int.

giffels replied:
I have added a convert_to function handling the conversion internally. So everything convertible to `int` is converted, strings that are not convertible are still handled as strings. For example the unique name of the drone.

giffels commented:
> Just wondering: Why not go for bytes directly? As far as I can tell, SLURM supports that as well.
> 
> Bytes never can be fractional, and it frees maintainers from having to look up whether `mb` is MB or MiB.

The only reason coming to my mind right now justifing MB is human readability in log files. SLURM is following standards `mb` for MB and `mib` MiB, together with using the factor `1000` in the code, that should clear enough.

However, I do not insists on it.

HerrHorizontal reviewed (commented):
Please check inline comment.

giffels commented:
> > Just wondering: Why not go for bytes directly? As far as I can tell, SLURM supports that as well.
> > Bytes never can be fractional, and it frees maintainers from having to look up whether `mb` is MB or MiB.
> 
> The only reason coming to my mind right now justifing MB is human readability in log files. SLURM is following standards `mb` for MB and `mib` MiB, together with using the factor `1000` in the code, that should be clear enough.
> 
> However, I do not insists on it.

How do you specify memory in bytes in Slurm. The default is MB 

```
sbatch --ntasks=1 --time=10 --mem=1 -p single test.sh
```

gives you 1 Megabyte.

```
sbatch --ntasks=1 --time=10 --mem=2000b -p single test.sh
```

does not work at all. @maxfischer2781 this could be a show stopper to use bytes.

HerrHorizontal reviewed (dismissed):
:+1:

maxfischer2781 commented:
> > ...
> 
> How do you specify memory in bytes in Slurm. The default is MB
> 
> ```
> sbatch --ntasks=1 --time=10 --mem=1 -p single test.sh
> ```
> 
> gives you 1 Megabyte.
> 
> ```
> sbatch --ntasks=1 --time=10 --mem=2000b -p single test.sh
> ```
> 
> does not work at all. @maxfischer2781 this could be a show stopper to use bytes.

That's weird. According to [the source](https://github.com/SchedMD/slurm/blob/3e7b49a16cc7403245a2d8d7bd1f2b0df3228128/src/common/slurm_protocol_defs.c#L6022) I would have expected the default to be bytes. However, [the sbatch docs](https://slurm.schedmd.com/sbatch.html) confirm that MB is the default and don't list a unit for raw bytes.

Anyway, if specifying bytes isn't straightforward, I don't see any advantage in doing so. MB resolution should be sufficient. Thanks for giving it a try.

olifre commented:
> That's weird. According to [the source](https://github.com/SchedMD/slurm/blob/3e7b49a16cc7403245a2d8d7bd1f2b0df3228128/src/common/slurm_protocol_defs.c#L6022) I would have expected the default to be bytes. However, [the sbatch docs](https://slurm.schedmd.com/sbatch.html) confirm that MB is the default and don't list a unit for raw bytes.

This made me check the sources again. It seems there are several places in which a unit conversion is performed independently :cry:. 
The parsing of the `--mem` option seems to happen here:
https://github.com/SchedMD/slurm/blob/2b38cb1cbe9fb76bee97bfdeb554b02bc9ff3142/src/common/slurm_opt.c#L2588
and this relies upon the `str_to_mbytes` function:
https://github.com/SchedMD/slurm/blob/15f2f073287aff5ebc150bf2eb2becb79f93c88d/src/common/proc_args.c#L404

This, in turn, does not allow bytes, and uses `1024` multipliers. Apparently, they also had a second function doing this for the command line tools, which could be switched to default to `G` units:
https://github.com/SchedMD/slurm/blob/50353e892e0ab52a43dc85d5cdc0fbcd8a97bf1d/src/common/proc_args.c#L511

So I have to revise my earlier comment: It seems that for the command line tools, `1024` multipliers are always used. Only the SLURM on-the-wire protocol then behaves as I linked earlier. Pretty confusing.

RHofsaess reviewed (dismissed):
From my (short) research, I think the suggested solution of Manuel (using mb) is the best and can be approved.

giffels commented:
> > That's weird. According to [the source](https://github.com/SchedMD/slurm/blob/3e7b49a16cc7403245a2d8d7bd1f2b0df3228128/src/common/slurm_protocol_defs.c#L6022) I would have expected the default to be bytes. However, [the sbatch docs](https://slurm.schedmd.com/sbatch.html) confirm that MB is the default and don't list a unit for raw bytes.
> 
> This made me check the sources again. It seems there are several places in which a unit conversion is performed independently 😢.
> The parsing of the `--mem` option seems to happen here:
> https://github.com/SchedMD/slurm/blob/2b38cb1cbe9fb76bee97bfdeb554b02bc9ff3142/src/common/slurm_opt.c#L2588
> and this relies upon the `str_to_mbytes` function:
> https://github.com/SchedMD/slurm/blob/15f2f073287aff5ebc150bf2eb2becb79f93c88d/src/common/proc_args.c#L404
> 
> This, in turn, does not allow bytes, and uses `1024` multipliers. Apparently, they also had a second function doing this for the command line tools, which could be switched to default to `G` units in the past, that seems gone / changed now:
> https://github.com/SchedMD/slurm/blob/50353e892e0ab52a43dc85d5cdc0fbcd8a97bf1d/src/common/proc_args.c#L511
> 
> So I have to revise my earlier comment: It seems that for the command line tools, `1024` multipliers are always used. Only the SLURM on-the-wire protocol then behaves as I linked earlier. Pretty confusing.

Thanks @olifre for investigating. I have changed the code accordingly and added a comment about the unit conversion in SLURM. Now, it should be consistent

maxfischer2781 commented on tardis/utilities/utils.py:
```diff
@@ -131,3 +131,10 @@ def submit_cmd_option_formatter(options: AttributeDict) -> str:
     return option_string.strip()
+
+
+def convert_to(value: Any, convert_to_type: Type, default: Any = None):
+    try:
+        return convert_to_type(value)
+    except ValueError:
+        return default
```
Should we actually go with a default ``None`` in any case here? I imagine this could hide some errors.

A sentinel could be used here to differentiate between "no default" and "default of ``None``".
```suggestion
sentinel = object()

def convert_to(value: Any, convert_to_type: Type, default: Any = sentinel):
    try:
        return convert_to_type(value)
    except ValueError:
        if default is sentinel:
            raise
        return default
```

maxfischer2781 replied:
Alternatively with naive typing:
```suggestion
from typing import TypeVar, Callable
T = TypeVar("T")
sentinel = object()

def convert_to(value: Any, convert_to_type: Callable[[Any], T], default: Any = sentinel) -> T:
    try:
        return convert_to_type(value)
    except ValueError:
        return default
```
The ``Callable`` is a bit more general than just types. Handling the ``default`` properly needs some boilerplate ``@overload``s, though, not sure if we want those.

maxfischer2781 reviewed (commented):
Traceback, most recent complaint last:
    File "tardis/utilities/utils.py", line 136, in Hackathon
Expected std::Review<Approve>, got std::Review<Comment>

maxfischer2781 reviewed (approved):
Approved be thee!

maxfischer2781 reviewed (approved):
Roses are red, violets are blue, this PR is approved now too!

RHofsaess reviewed (approved):
:+1:
